File: 101/symmetric-tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func isSymmetric(root *TreeNode) bool {
	if nil == root {
		return true
	}
	var isMirror func(*TreeNode, *TreeNode) bool
	isMirror = func(root1 *TreeNode, root2 *TreeNode) bool {
		if nil == root1 && nil == root2 {
			return true
		} else if nil == root1 || nil == root2 {
			return false
		} else {
			return (root1.Val == root2.Val) && isMirror(root1.Left, root2.Right) && isMirror(root1.Right, root2.Left)
		}
	}
	return isMirror(root.Left, root.Right)
}

File: 101/symmetric-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Solution:
    def isSymmetric(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        if root is None:
            return True
        
        def isMirror(root1, root2):
            if root1 is None and root2 is None:
                return True
            if root1 is None or root2 is None:
                return False
            return (root1.val == root2.val) and isMirror(root1.left, root2.right) and isMirror(root1.right, root2.left)
        return isMirror(root.left, root.right)
    